docs/guide/testing-types.md

```diff
@@ -0,0 +1,81 @@
+---
+title: Testing Types | Guide
+---
+
+# Testing Types
+
+Vitest allows you to write tests for your types, using `expectTypeOf` or `assertType` syntaxes. By default all tests inside `*.test-d.ts` files are considered type tests, but you can change it with [`typecheck.include`](/config/#typecheck-include) config option.
+
+Under the hood Vitest calls `tsc` or `vue-tsc`, depending on your config, and parses results. Vitest will also print out type errors in your source code, if it finds any. You can disable it with [`typecheck.ignoreSourceErrors`](/config/#typecheck-ignoresourceerrors) config option.
+
+Keep in mind that Vitest doesn't run or compile these files, they are only statically analyzed by the compiler, and because of that you cannot use any dynamic statements. Meaning, you cannot use dynamic test names, and `test.each`, `test.runIf`, `test.skipIf`, `test.each`, `test.concurrent` APIs. But you can use other APIs, like `test`, `describe`, `.only`, `.skip` and `.todo`.
+
+Using CLI flags, like `--allowOnly` and `-t` are also supported for type checking.
+
+```ts
+import { assertType, expectTypeOf } from 'vitest'
+import { mount } from './mount.js'
+
+test('my types work properly', () => {
+  expectTypeOf(mount).toBeFunction()
+  expectTypeOf(mount).parameter(0).toMatchTypeOf<{ name: string }>()
+
+  // @ts-expect-error name is a string
+  assertType(mount({ name: 42 }))
+})
+```
+
+Any type error triggered inside a test file will be treated as a test error, so you can use any type trick you want to test types of your project.
+
+You can see a list of possible matchers in [API section](/api/#expecttypeof).
+
+## Reading Errors
+
+If you are using `expectTypeOf` API, you might notice hard to read errors or unexpected:
+
+```ts
+expectTypeOf(1).toEqualTypeOf<string>()
+//             ^^^^^^^^^^^^^^^^^^^^^^
+// index-c3943160.d.ts(90, 20): Arguments for the rest parameter 'MISMATCH' were not provided.
+```
+
+This is due to how [`expect-type`](https://github.com/mmkal/expect-type) handles type errors.
+
+Unfortunately, TypeScript doesn't provide type metadata without patching, so we cannot provide useful error messages at this point, but there are <a href="https://github.com/microsoft/TypeScript/pull/40468" tatger="_blank">works in TypeScript project</a> to fix this. If you want better messages, please, ask TypeScript team to have a look at mentioned PR.
+
+If you find it hard working with `expectTypeOf` API and figuring out errors, you can always use more simple `asserType` API:
+
+```ts
+const answer = 42
+
+assertType<number>(answer)
+// @ts-expect-error answer is not a string
+assertType<string>(answer)
```

Comment:
Just to add my two cents: in type testing `// @ts-expect-error` comment should be considered an anti pattern, because it hides any kind of error. For instance, a test is a type is passing:

```ts
// @ts-expect-error answer is not a string
 assertType<string>(answr)
```

Or, for example, `assertType` can be renamed or removed, but the test will be perfectly passing. Ouch... Perhaps executing the file could be a solution? `expect-type` has no runtime, but Vitest has one.

Reply:
Yeah, but if you don't use this pattern (which is not a recommended way to do it, we encourage `expect-type`), then we just spent a lot of time by running code that could've been just statically analysed.

If you want, you can always add both runtime and type tests to both `test.include` and `test.typecheck.include`

Reply:
Ah right! Could be very useful note to add. (Ignore me, if it is already here and I missed it)

Reply:
Added it to the docs, thanks for your input 👍🏻